graphviz.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python
+
+"""
+Pandoc filter to process code blocks with class "graphviz" into
+graphviz-generated images.
+
+Needs pygraphviz
+"""
+
+import os
+import sys
+
+import pygraphviz
+import string
+
+from pandocfilters import toJSONFilter, Para, Image, get_filename4code, get_caption, get_extension, get_value
+
+def graphviz(key, value, format, _):
+    if key == 'CodeBlock':
+        [[ident, classes, keyvals], code] = value
+        if "graphviz" in classes:
+            caption, typef, keyvals = get_caption(keyvals)
+            prog, keyvals = get_value(keyvals, u"prog", u"dot")
+            filetype = get_extension(format, "png", html="png", latex="pdf")
+            dest = get_filename4code("graphviz", code, filetype)
+            filename = ""
+            for a,fname in keyvals:
+                if(a == "filename"):
+                    valid_chars = "-_ %s%s" % (string.ascii_letters, string.digits)
+                    fname = ''.join(x for x in fname if x in valid_chars)
+                    if fname != "":
+                        filename = "graphviz-images/"+fname+"."+filetype
+                        dest = filename
+            if (filename != "") or (not os.path.isfile(dest)):
+                g = pygraphviz.AGraph(string=code)
+                g.layout()
+                g.draw(dest, prog=prog)
+                sys.stderr.write('Created image ' + dest + '\n')
```

Comment:
I think we can remove the directory `graphviz-images`, because it will be created automatically, right?

Reply:
If the markdown is directed accessed the image can be seen from the .jpg

![image](https://user-images.githubusercontent.com/17395762/57983613-77490500-7a2a-11e9-831f-92cf3c6cea9d.png)

Reply:
We also use this directory for caching and direct markdown visualization... I think it's important for github contributors, without access to compiling the pdf. This is an advantage of pandoc over pure LaTeX format.